clients/pkg/promtail/wal/watcher.go

```diff
@@ -0,0 +1,305 @@
+package wal
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/record"
+	"github.com/prometheus/prometheus/tsdb/wlog"
+
+	"github.com/grafana/loki/pkg/ingester/wal"
+)
+
+const (
+	readPeriod         = 10 * time.Millisecond
+	segmentCheckPeriod = 100 * time.Millisecond
+
+	// debug flag used for developing and testing the watcher code. Using this instead of level.Debug to avoid checking
+	// the log level inwards into the logger code, and just making the compiler omit this code.
+	debug = false
+)
+
+// Based in the implementation of prometheus WAL watcher
+// https://github.com/prometheus/prometheus/blob/main/tsdb/wlog/watcher.go. Includes some changes to make it suitable
+// for log WAL entries, but also, the writeTo surface has been implemented according to the actions necessary for
+// Promtail's WAL.
+
+// Reader is a dependency interface to inject generic WAL readers into the Watcher.
+type Reader interface {
+	Next() bool
+	Err() error
+	Record() []byte
+}
+
+// WriteCleanup is responsible for cleaning up resources used in the process of reading the WAL.
+type WriteCleanup interface {
+	// SeriesReset is called to notify that segments have been deleted. The argument of the call
+	// means that all segments with a number lower or equal than segmentNum are safe to be reclaimed.
+	SeriesReset(segmentNum int)
+}
+
+// WriteTo is responsible for doing the necessary work to process both series and entries while the Watcher
+// is reading / tailing segments. Note that StoreSeries and SeriesReset might be called concurrently.
+//
+// Based on https://github.com/prometheus/prometheus/blob/main/tsdb/wlog/watcher.go#L46
+type WriteTo interface {
+	WriteCleanup
+
+	// StoreSeries is called when series are found in WAL entries by the watcher, alongside with the segmentNum they were
```

Comment:
Just one nit: have we considered going with the Promtail terminology (and call them streams maybe?) instead of keeping close to what Prometheus does?